### ISSUE TITLE
[data grid] axe 'empty-table-header" violation when testing with actioncolumn

### ISSUE THREAD

**Issue by a-vogt** [mui/mui-x#20203]
### Steps to reproduce

Steps:
1. Clone [repo ](https://github.com/a-vogt/mui-issue-actioncolumns-axe)
2. npm install
3. npm run test

### Current behavior

We use the action column as described under actions [here ](https://mui.com/x/react-data-grid/column-definition/#special-properties) and get the following error when we run axe test:

![Image](https://github.com/user-attachments/assets/6622e612-9453-4b6b-b1e3-fe045a9edf0c)

### Expected behavior

no violation

### Context

Is there anything we can do to fix this error/be accessible compliant, or is this a bug that needs to be fixed on your end?

**Search keywords**: data grid, axe, test, empty-table-header, action column, a11y, accessibility

**Order ID**: 116257

**Comment by MBilalShafi** [comment]
The "actions" column doesn't have a `headerName` by default. You can add the `headerName` property to column definition of the actions column to add a title which should hopefully fix the issue with your a11y rules.

I updated the demo in this example for reference: https://stackblitz.com/edit/39gzi1vn?file=src%2FDemo.tsx

Let me know if it fixes your issue.

**Comment by a-vogt** [comment]
Thank you for your quick response and solution idea.

I set `headerName` and it passed the test, but unfortunately that's not a solution for me because I don't want to display a title for this column.
I tried several options to set a string in headerName and still hide the title:
```
headerName: ‘’,
headerName: ‘ ’,
renderHeader: () => null,
renderHeader: () => <></>,
```

Empty string, single space or `renderHeader` still violate the test [“Table header text should not be empty (empty-table-header)”].

I would prefer not to solve this via CSS/SX, as shown here:
 
```
.MuiDataGrid-columnHeader[data-field=“string”] .MuiDataGrid-columnHeaderTitle {
  display: none;
}
```

**Is there a way to set `headerName` and still hide the title?**

Thanks in advance.

**Comment by MBilalShafi** [comment]
> Is there a way to set headerName and still hide the title?

I'm curious about the rationale around it. 

Your accessibility guideline explicitly states that “_Table header should not be empty._”

From my perspective, there are two valid approaches: either disable the rule if you really need an empty header, or provide a meaningful header if you want to comply with the rule. Trying to hide the header text while keeping the rule enabled might make tests pass but would still go against project accessibility guidelines.

I’d appreciate any clarification on whether that understanding is correct.

**Comment by a-vogt** [comment]
You understood correctly.

Of course, I can disable the rule
`‘empty-table-header’: { enabled: false }`,
but there are obviously good reasons not to do so.

The question is whether an (action) column without a title must generally comply with accessibility guidelines, e.g., by setting aria-label or similar.

Unfortunately, setting a title is somewhat unattractive in this case because the row could be clickable and only an indicator is visible in the action column.

Perhaps I am overlooking something, but I would like the system to automatically set something like this for an empty column title in order to be accessible.

**Comment by MBilalShafi** [comment]
The `empty-table-header` rule explicitly rejects `aria-label` as a solution.

According to [Deque University's documentation](https://dequeuniversity.com/rules/axe/4.10/empty-table-header), table headers must contain **discernible text** (visible or visually hidden), not just ARIA attributes. Using only `aria-label` fails because:
  - Screen reader users get the label
  - Sighted users (including those with low vision, cognitive disabilities, or keyboard navigation) cannot perceive the
  column's purpose
  - This creates an inequitable experience

As a workaround, you could use **visually hidden text** in the column header.

```jsx
{
  field: 'actions',
  renderHeader: () => (
    <>
      <span className="visually-hidden">Actions</span>
      {/* Optional visible icon */}
    </>
  ),
  // ... rest of column config
}


// With CSS
.visually-hidden {
  position: absolute;
  left: -10000px;
  width: 1px;
  height: 1px;
  overflow: hidden;
}
```

Updated Stackblitz: https://stackblitz.com/edit/39gzi1vn?file=src%2FDemo.tsx

This should fix the `empty-table-header` accessibility issue you are encountering.

Now, I'm not sure if axe-core is a universally accepted accessibility standard and if we should fix this internally too, but until then you could use the solution.

**Comment by a-vogt** [comment]
Sorry for the late reply and thank you for the suggested solution.

**Since we already do this in other components to support screen readers only, you are welcome to mark your suggestion as an accepted solution.**

Perhaps you could take another look at it internally and consider, when you have time, whether such points for screen readers and accessibility should be set elsewhere, rather than having to write another CSS class.
